nautobot/utilities/filters.py

```diff
@@ -286,6 +286,141 @@ def filter(self, qs, value):
+        """
+        Given a mapping of `filter_predicates` and a `value`, return a `Q` object for 2-tuple of
+        predicate=value.
+        """
+        query = models.Q()
+        for field_name, lookup_info in filter_predicates.items():
+            if isinstance(lookup_info, str):
+                lookup_expr = lookup_info
+                preprocessor = str.strip
+            elif isinstance(lookup_info, dict):
+                lookup_expr = lookup_info["lookup_expr"]
+                preprocessor = lookup_info.get("preprocessor")
+                if not callable(preprocessor):
+                    raise TypeError("Preprocessor {preprocessor} must be callable!")
+            else:
+                raise TypeError(f"Predicate value must be a str or a dict! Got: {type(lookup_info)}")
+
+            # Try to preprocess the value or skip creating a predicate for it. In the event we try
+            # to cast a value to an invalid type (e.g. `int("foo")` or `dict(42)`), ensure this
+            # predicate is not included in the query.
+            try:
+                value = preprocessor(value)
+            except (TypeError, ValueError):
+                continue
+
+            predicate = {f"{field_name}__{lookup_expr}": value}
+            query |= models.Q(**predicate)
+
+        # Return this for later use (such as introspection or debugging)
+        return query
+
+    def filter(self, qs, value):
+        if value in EMPTY_VALUES:
+            return qs
+
+        # Evaluate the query and stash it for later use (such as introspection or debugging)
+        self.query = self.generate_query(self.filter_predicates, value)
+
+        qs = self.get_method(qs)(self.query)
+        return qs.distinct()
+
+
+class SearchFilter(MappedPredicatesFilterMixin, django_filters.CharFilter):
+    """
+    Provide a search filter for use on filtersets as the `q=` parameter.
+
+    A mapping of filter predicates (field_name: lookup_expr) must be provided to the filter when
+    declared on a filterset. This mapping is used to construct a `Q` query to filter based on the
+    provided predicates.
+
+    This filter should be used instead of declaring a `q` filter and pointing it to a `search()`
+    method on a filterset.
+
+    Example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "comments": "icontains",
+                "name": "icontains",
+                "id": "iexact",
+            },
+        )
+
+    Optionally you may also provide a callable to use a a preprocessor for the filter predicate by
+    providing the value as a nested dict with "lookup_expr" and "preprocessor" keys. For example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "asn": {
+                    "lookup_expr": "exact",
+                    "preprocessor": int,
+            },
+        )
+
+    This tells the filter to try to cast `asn` to an `int`. If it fails, this
+    predicate will be skipped.
```

Comment:
Personally I'd rather not duplicate this information across two different docstrings - makes it more likely one or the other will not get updated over time.

Reply:
Same.